### DISCUSSION TITLE
Hardware acceleration with Hastlayer - consultant wanted!

### DISCUSSION THREAD

**Discussion by Piedone** [SixLabors/ImageSharp#1352] [Q&A]
Hi everyone,

we at [Lombiq](https://lombiq.com/) run a pretty unique project called [Hastlayer](https://hastlayer.com/): Hastlayer can automatically transform compatible .NET code into a hardware implementation, making the result much faster and consume much less power (along the lines of 5-10x in suitable cases). We'd like to implement a Hastlayer-accelerated ImageSharp PoC and are looking for a developer experienced with the internals of ImageSharp to help us with it. Would you like to work with us? (And I mean paid consulting of course.)

The PoC, if successful, would serve as a demonstration in web apps (like [Orchard Core](https://www.orchardcore.net/) apps, something that we're actually focused on as a company), as well as advanced scenarios like running image processing tasks for space industry use-cases, including ground segment and on board of a satellite (yes, we actually want to put Hastlayer on a satellite, and it's not as far-fetched as you may think).

Hastlayer is not a silver bullet and can only help if the algorithm in question is massively parallelizable (using hundreds of `Task`s) and compute or memory-bound, see details [here](https://github.com/Lombiq/Hastlayer-SDK/blob/client/Docs/WorkingWithHastlayer.md#performance-optimizing-your-code). Also, the code needs to be compatible with what Hastlayer supports, which currently is elaborated [here](https://github.com/Lombiq/Hastlayer-SDK/blob/client/Docs/WorkingWithHastlayer.md#writing-hastlayer-compatible-net-code). There is a simple but well-performing [image processing sample](https://github.com/Lombiq/Hastlayer-SDK/blob/client/Samples/Hast.Samples.SampleAssembly/ImageContrastModifier.cs) already.

BTW all of this happens with [FPGAs](https://en.wikipedia.org/wiki/Field-programmable_gate_array) which are reconfigurable chips that can mimic other logic hardware. You can think of them as kind of an accelerator card in this use-case. So no need to actually manufacture the hardware generated by Hastlayer though that would also be a possibility :). FPGAs are available in the cloud too, like in AWS and Azure.

What do you think? If you're interested, please let me know here or send me an e-mail to my address under my profile.

**Comment by antonfirsov** [comment]
Hey @Piedone nice to see you here again! :)

The hardest part is that the vast majority of our codebase doesn't play well with those limitations, the most painful being the lack of support for  IEEE-754 `float`-s. When we last discussed this topic, the most realistic POC idea we were able to pick was to to implement a a very basic variant of [`ResizeProcessor`](https://github.com/SixLabors/ImageSharp/blob/master/src/ImageSharp/Processing/Processors/Transforms/Resize/ResizeProcessor%7BTPixel%7D.cs) on top of Hastlayer. I think this still stands.

As of my personal involvement, I really struggle to find time these days, even for the basic ImageSharp contributions, so I can't promise anything 😞 

@Sergio0694 likes to try crazy things, maybe you can convince him to help you :wink:

**Reply by Piedone** [reply]
Hi again Anton!

Yes, while standard `float`s (or `double` or `Half`) are not available, calculating with fractions is: Either with a 64b fixed-point type or with 8-64b [posits](https://posithub.org/). However, neither of these are particularly efficient.

`ResizeProcessor` still looks like a good candidate!

**Reply by antonfirsov** [reply]
The big challenge here is that this problem isn't really specific to ImageSharp.

If you aim for good adoption rates, you need good portability, but existing computation-heavy .NET libraries (which would be ideally a good candidates for "Hastlayerization") will very likely use float stuff.

If it's really hard/not worth to implement IEEE-754 on FPGA, I would at least aim for finding an easy way to manually (or semi-automatically) port existing float-based codebases to posits, or to some fixed-point type accelerated by Hastlayer. Otherwise you will always end up implementing an alternative path just like with our Resizer. (Which means that you'll always be forced to have a good understanding of the underlying algorithms.)

**Reply by Piedone** [reply]
Yeah, that's some good feedback, thank you. On certain boards, IEEE-754 floating-point is actually available in hard DSPs (otherwise you have to configure a portion of the FPGA fabric to serve as a floating-point ALU, which is possible but inefficient) which we could make available. IEEE-754 floats are not guaranteed to provide the exact same results between different hardware platforms though so that's a challenge.

**Reply by antonfirsov** [reply]
> IEEE-754 floats are not guaranteed to provide the exact same results between different hardware platforms though so that's a challenge.

A lesson every graphics programmer learns very early 😉

I would put this into the "not good, not terrible" category. ImageSharp does not produce identical results across platforms, 99.99% of our users are fine with this, unless results look visually good.

**Reply by Piedone** [reply]
So anybody interested? @Sergio0694? Or @brianpopow, @tocsoft (taking a look at recent top contributors)? And obviously @JimBobSquarePants but I guess you have enough on your plate already.

**Reply by brianpopow** [reply]
@Piedone i do find the idea interesting, but i barely find time to do my own tasks for ImageSharp at the moment. So taking another task on top of that is just not feasible for me.

**Reply by Piedone** [reply]
I see, sorry to hear that.
